Packet-03: add assignments 1 through 4

aspsk opened the pull request:
Implement assignments and describe them in the README.org file.

aspsk commented:
Sorry for the force-push. Did two changes.

First, fixed the `"xdp_redirect"` program (as it failed to load after I removed the real code to form the assignment).

Second, added a note that we can't (in general) use `XDP_REDIRECT` in the native mode for `veth` devices, as a target interface receiving side expects a connected XDP program (see https://github.com/torvalds/linux/commit/af87a3aa1b5f397a2f5c99b97b000943c5177da7)

tohojo commented:
Thanks for doing this! I'll see if I can't get around to reviewing this sometime this week :)

tohojo commented:
Right, some comments:

- The "old code" you are removing in your first commit is actually the solution
  for the previous assignment. We should rather keep this at the end of the
  file, like we do in the other assignments. Also, please add a solution to this
  new assignment somewhere; maybe just create a packet04-solution directory for
  now (we can then rename it when we figure out what the next assignment should
  be).

- In the previous assignments we've split up the README text into "concepts you
  are going to learn" and "assignment text". Here, you do have the explanations,
  but they are part of the assignment text; please split them like in the other
  assignments.

- The checksum update could  probably be done using the bpf_csum_diff() helper,
  so let's add a description of that as well.

- We also need to explain why one should use the redirect_map() helper instead
  of the plain redirect() helper.

- The point about needing a running XDP program on the interface that is being
  targeted by the redirect needs more explanation, as this is also something
  people are likely to run into on physical hardware. So this point should be
  expanded a bit upon in the explanation (once that is moved to the "things you
  will learn section, as per above). Also, please add a note that we are working
  on fixing this upstream :)

- Could we keep using the same loader from the previous assignment, and just
  create a separate program to populate (and read from) the map? This could
  serve a pedagogical purpose as well, by teaching the user that they will need
  to create their own protocol (and UI) when using maps, whereas the
  fib lookup helper will let them keep using the kernel builtins.

- Currently, assignment 3 is not much work, while assignment 2 is quite a bit.
  Maybe a better split would be to just hard-code the MAC addresses in
  assignment 2, and then move the map-based lookup to assignment 3?

- On the subject of the map lookups, I think the current map API feels a bit
  contrived. Why do four map lookups, then memcmp() on the values, when we can
  just turn the memcmp() values into the map key? And we don't really need to
  use the source MAC in the lookup, do we?

aspsk commented:
Thanks for the comments, all are pretty reasonable! I will work on addressing them.

Several comments/questions:

1. If we want to spread complexity between the 2-nd and 3-rd assignments, then I propose to use `bpf_redirect` and all parameters hardcoded in the assignment 2, and `bpf_redirect_map` and a proper program to load/read maps in the assignment 3.

2. I will extract commits not directly related to this PR into a separate one. (Simplify `xdp_stats_record_action`, fix `--legacy-ip`, update `bpf_helpers.h`).

What do you think?

tohojo commented:
Anton Protopopov <notifications@github.com> writes:

> Thanks for the comments, all are pretty reasonable! I will work on
> addressing them.

Great, thanks!

> Several comments/questions:
>
> 1. If we want to spread complexity between the 2-nd and 3-rd
> assignments, then I propose to use `bpf_redirect` and all parameters
> hardcoded in the assignment 2, and `bpf_redirect_map` and a proper
> program to load/read maps in the assignment 3.

Hmm, as long as we very clearly state that people really should be using
the redirect_map() helper for anything outside of the tutorial, I guess
that is fine. There is quite a significant difference in performance
between the two...

> 2. I will extract commits not directly related to this PR into a
> separate one. (Simplify `xdp_stats_record_action`, fix `--legacy-ip`,
> update `bpf_helpers.h`).

SGTM :)

aspsk commented:
Hi @tohojo

Updated the PR

tohojo commented on packet-solutions/xdp_prog_kern.c:
```diff
@@ -0,0 +1,419 @@
+		goto out;
+	}
+
+	icmp_type = parse_icmphdr(&nh, data_end, &icmphdr);
```
Alternatively, create a new helper that makes it explicit that you are parsing the common fields...

aspsk replied:
You're right. In the solution below I am also using the result struct to patch the icmp/icmpv6 header, also could be non-obvious. I will rewrite this piece.

tohojo commented on packet03-redirecting/README.org:
```diff
@@ -5,49 +5,277 @@
+won't deliver redirected/retransmitted XDP frames unless there is an XDP
+program attached to the receiving side of the target =veth= interface. Physical
+hardware will likely behave the same. XDP maintainers are currently working on
+fixing this behaviour upstream.
```
I think we should still use the native mode in later assignments; it may be a while before this is fixed upstream, and in the meantime it's important to show a user how they are going to get things working on real hardware.

So explain that loading an XDP program on the other interface is needed, and make it part of assignment 3 and 4 to actually do this.

aspsk replied:
OK, sure. I also will add a link to this talk: https://www.netdevconf.org/0x13/session.html?talk-veth-xdp

tohojo commented on packet03-redirecting/README.org:
```diff
@@ -5,49 +5,277 @@
+supported and load the XDP program:
+#+begin_src sh
+$ t setup -n test --legacy-ip
+$ sudo ./xdp_loader -d test --skb -F --progsec xdp_icmp_echo
```
It should be possible to just use `t load` here, no?

tohojo commented on packet03-redirecting/README.org:
```diff
@@ -5,49 +5,277 @@
+Setup the two environments, patch the =xdp_redirect= program accordingly, and
+attach it to the =right= interface.  To test load the program, enter the right
+environment, and ping the inner interface of the left environment (your IPv6
+address will be different):
```
s/will be/may be/

aspsk replied:
+1

tohojo commented on packet03-redirecting/README.org:
```diff
@@ -5,49 +5,277 @@
+#$ t setup -n left
+#$ t setup -n right
+#$ sudo ./xdp_loader_redirect --progsec xdp_redirect --skb -F -d right -r left
+##+end_src
```
Looks like the commented-out bit is just left over from earlier and should be removed?

aspsk replied:
Yes, thanks!

tohojo commented on packet03-redirecting/README.org:
```diff
@@ -5,49 +5,277 @@
+example from the Linux kernel, but was patched to update statistics as in other
+examples and to check all the return values from the =bpf_fib_lookup()= function.
+Don't forget to update the =xdp_prog_user.c= program to setup 1-1 mapping
+between virtual ports and network devices.
```
This should probably be explained in a bit more detail. Maybe expand the explanation of bpf_fib_lookup() in the previous section, to explain that it returns ifindexes, but needs virtual ports?

aspsk replied:
ok

tohojo commented on packet03-redirecting/xdp_prog_kern.c:
```diff
@@ -7,124 +7,261 @@
 
-        /* Make a copy of the outer Ethernet header before we cut it off */
-        __builtin_memcpy(&eth_cpy, eth, sizeof(eth_cpy));
+	icmp_type = parse_icmphdr(&nh, data_end, &icmphdr);
```
Same comment about icmp header applies as in the solution, obviously :)

tohojo commented on packet03-redirecting/README.org:
```diff
@@ -5,49 +5,277 @@
+net.ipv6.conf.all.forwarding = 1
+$ sudo sysctl net.ipv4.ip_forward=1
+net.ipv4.ip_forward = 1
+#+end_src
```
I think we can just enable forwarding per-interface in the venv setup code in testenv.sh, can't we? I'd keep the note about forwarding needing to be enabled, but would rather not instruct users to make global changes to their machine just to run the tutorial...

aspsk replied:
Ok, per-interface forwarding is better. But I would rather do it as a part of solution. It is fun that we can redirect packets in the Assignments 2 and 3 without enabling forwarding on interfaces. What do you think?

tohojo replied:
Right, good point about demonstrating that we can redirect without forwarding being turned on. But in that case we should probably point that out somewhere ;)

aspsk replied:
Ah, I think that I mentioned that, but it was lost when I addressed your first series of comments

aspsk commented:
> This is looking really good, thanks! Still a few things to fix (see below), but I think we are quite close to where this can be merged! :)

Thanks for reviewing this!

> One general comment, in addition to the specific comments below: I'm OK with moving the solutions to a separate directory, but in that case, please put them in separate files and name the files after the lesson they are a solution to (e.g., 'packet03_solution_kern.c'). That way it is easier to find the solution for a previous assignment without inadvertently looking at the solution for the next one as well...

This makes sense, I will split solutions into separate files. Also, not related to this PR, but am I right that a solution to the Assignment 1 of the packet-02 is missing?

aspsk commented on packet-solutions/xdp_prog_kern.c:
```diff
@@ -0,0 +1,419 @@
+	ipv6->daddr = tmp;
+}
+
+static __always_inline void swap_src_dst_ipv4(struct iphdr *iphdr)
```
ok (and for all the changes requested above)

aspsk commented on packet-solutions/xdp_prog_kern.c:
```diff
@@ -0,0 +1,419 @@
+	m0 = * (__u16 *) icmphdr;
+	icmphdr->type = echo_reply;
+	m1 = * (__u16 *) icmphdr;
+	icmphdr->checksum = ~(csum16_add(csum16_add(~icmphdr->checksum, ~m0), m1));
```
Sure, I will replace this with `bpf_csum_diff`. I wanted to use a hand-made implementation so that we can precisely see that the new sum is just `~(~sum +  ~m0 +  m1)`

tohojo commented:
> This  makes sense, I will split solutions into separate files. Also, not  related to this PR, but am I right that a solution to the Assignment 1  of the packet-02 is missing?

You are right in the sense that there is no separate program that just calls vlan_tag_pop() - the "solution" is just the implementation of that function...

aspsk commented:
> > This  makes sense, I will split solutions into separate files. Also, not  related to this PR, but am I right that a solution to the Assignment 1  of the packet-02 is missing?
> 
> You are right in the sense that there is no separate program that just calls vlan_tag_pop() - the "solution" is just the implementation of that function...

I believe that the first assignment is "Assignment 1: Rewrite port numbers" which proposes to patch UDP ports of incoming packages

tohojo commented:
Doh, you're right of course! Sorry, I mentally got the '2' from 'packet-02' loaded into my "assignment number" register when looking this up :P

And yeah, you are also quite right, seems I forgot to include a solution for assignment 1. If you want to, feel free to add it, otherwise I'll do that after we've merged this :)

irevoire commented:
Just tried this as someone learning XDP and it's great!
Thanks a lot for your works! :tada:

irevoire commented:
Also in the readme you talk about the `csum_replace2`, which seams easier to use than `bpf_csum_diff`. 
But this function is not defined in this directory, I think it would be great to copy paste it somewhere under the `common` directory.
Or directly in the `xdp_prog_kern.c` since right now it's the only time we need it in all the exercises.

aspsk commented:
> Also in the readme you talk about the `csum_replace2`, which seams easier to use than `bpf_csum_diff`.
> But this function is not defined in this directory, I think it would be great to copy paste it somewhere under the `common` directory.
> Or directly in the `xdp_prog_kern.c` since right now it's the only time we need it in all the exercises.

As Toke asked, I will use the `bpf_csum_diff` function in the solution, as this will be more generic and can be used in other cases. However, I will leave the alternative "handmade" solution as a comment right next.

aspsk commented:
Hi @tohojo I've updated the PR. One thing which I didn't do is that I didn't split solutions into separate files. Otherwise I think that I've addressed all your comments.  Please review without this last change—I will do this in a consequent PR and also will add a solution for the missing assignment.

aspsk commented:
@tohojo One thing you may note is that I am still using the `sysctl net.ipv6.conf.all.forwarding=1` for the `bpf_fib_lookup` assignment.  It turned out that there is a bug for the IPv6 case—it is not sufficient to enable IPv6 forwarding for particular devices, only for all. I've just sent a patch to fix this, but before it is applied we will need to enable IPv6 forwarding globally.

tohojo commented:
Noted, thanks! I'll try to get around to looking at this again ASAP, but
I'm travelling this week, so it may not be until next week...

tohojo commented:
Merged with a few fixes. Thanks a lot for keeping at this! :)
